File: Week_04/529/minesweeper2.go
package main

func updateBoard2(board [][]byte, click []int) [][]byte {
	row, column := len(board), len(board[0])
	x, y := click[0], click[1]
	if board[x][y] == 'M' {
		board[x][y] = 'X'
		return board
	}
	dx := [8]int{-1, -1, -1, 0, 0, 1, 1, 1}
	dy := [8]int{-1, 1, 0, 1, -1, 0, 1, -1}
	var dfs func(r, c int)
	dfs = func(r, c int) {
		count := 0
		for i := 0; i < 8; i++ {
			x, y := r+dx[i], c+dy[i]
			if x >= 0 && x < row && y >= 0 && y < column && board[x][y] == 'M' {
				count++
			}
		}
		if count > 0 {
			board[r][c] = byte('0' + count)
		} else {
			board[r][c] = 'B'
			for i := 0; i < 8; i++ {
				x, y := r+dx[i], c+dy[i]
				if x >= 0 && x < row && y >= 0 && y < column && board[x][y] == 'E' {
					dfs(x, y)
				}
			}
		}
	}
	dfs(x, y)
	return board
}

File: Week_04/529/minesweeper2.py
from typing import List


class Solution:
    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        x, y = click[0], click[1]
        row, column = len(board), len(board[0])
        if board[x][y] == 'M':
            board[x][y] = 'X'
            return board

        def dfs(r, c):
            count = 0
            for dx, dy in [(-1, -1), (-1, 1), (-1, 0), (1, -1), (1, 1), (1, 0), (0, 1), (0, -1)]:
                if 0 <= r + dx < row and 0 <= c + dy < column and board[r + dx][c + dy] == 'M':
                    count += 1
            if count > 0:
                board[r][c] = str(count)
            else:
                board[r][c] = 'B'
                for dx, dy in [(-1, -1), (-1, 1), (-1, 0), (1, -1), (1, 1), (1, 0), (0, 1), (0, -1)]:
                    if 0 <= r + dx < row and 0 <= c + dy < column and board[r + dx][c + dy] == 'E':
                        dfs(r + dx, c + dy)

        dfs(x, y)
        return board
